### PR TITLE
NT-1958: UX – Post a reply

### PR DESCRIPTION
**Pull request by sunday-okpoluaefe**
# 📲 What

UX – Post a reply

# 🤔 Why

- to handle how we post a reply within a thread!

# 🛠 How
- Refactored ```PostCommentData```
- Refactored apolloclient apii to handle replies
```
... parentId(comment.parent?.let { encodeRelayId(it)})

```

# 👀 See

| Before 🐛 |
```
 this.service.mutate(
                CreateCommentMutation.builder()
                    .parentId(comment.parentId)
                    .commentableId(comment.commentableId)
                    .clientMutationId(comment.clientMutationId)
                    .body(comment.body)
                    .build()
            )
```
| After 🦋 |

```
 this.service.mutate(
                CreateCommentMutation.builder()
                    .parentId(comment.parent?.let { encodeRelayId(it)})
                    .commentableId(comment.commentableId)
                    .clientMutationId(comment.clientMutationId)
                    .body(comment.body)
                    .build()
            )
```
| --- | --- |
|  |  |

# 📋 QA
- Add this snippet to the  ``` ThreadViewmodel ``` init block
```

 /*
             * This block should be removed as its meant to test reply to a comment
             * The block above should show the replies
             */
            commentData
                .switchMap {
                    this.apolloClient.createComment(
                        PostCommentData(
                            commentableId = it.commentableId ?: "",
                            body = "Just a comment reply",
                            clientMutationId = null,
                            parent = it.comment
                        )
                    )
                }
                .compose(bindToLifecycle())
                .subscribe(
                    {
                        val c = it
                    },
                    {
                        val e = it
                    }
                )
```
- Go to any backed project comment screen
- Select any comment with replies. A reply is automatically posted 

# Story 📖

https://kickstarter.atlassian.net/browse/NT-1958

### CODE REVIEW

File: app/src/main/java/com/kickstarter/viewmodels/ThreadViewModel.kt [hunk 1]
```diff
@@ -79,9 +81,36 @@ interface ThreadViewModel {
                     )
                 }
 
-            comment
+            /*
+             * This block should be removed as its meant to test reply to a comment
+             * The block above should show the replies
+             */
+            commentData
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/viewmodels/ThreadViewModel.kt:88]: I think you can remove this block and just add it on the Pull Request on the QA section for the rest of us to QA it. Once that's done I'll approve the PR, it seems ok to me 👍

Comment (sunday-okpoluaefe) [reply app/src/main/java/com/kickstarter/viewmodels/ThreadViewModel.kt]: done

File: app/src/main/java/com/kickstarter/viewmodels/CommentsViewHolderViewModel.kt [hunk 2]
```diff
@@ -266,15 +266,15 @@ interface CommentsViewHolderViewModel {
...
                     )
                 }
                 .map {
                     PostCommentData(
                         commentableId = it.first,
-                        body = it.second,
+                        body = it.second.body(),
                         clientMutationId = null,
-                        parentId = null
+                        parent = it.second?.parentId()?.let { it1 -> it.second.toBuilder().id(it1).build() }
```
Comment (leighdouglas) [inline app/src/main/java/com/kickstarter/viewmodels/CommentsViewHolderViewModel.kt:277]: nit: can we rename `it1` to `id`?

File: app/src/main/java/com/kickstarter/viewmodels/ThreadViewModel.kt [hunk 3]
```diff
@@ -68,14 +67,14 @@ interface ThreadViewModel {
                 .compose(bindToLifecycle())
                 .subscribe(this.focusOnCompose)
 
-            val commentEnvelope = getCommentFromIntent()
+            val commentEnvelope = getCommentCardDataFromIntent()
                 .switchMap {
-                    this.apolloClient.getRepliesForComment(it)
+                    it.comment?.let { it1 -> this.apolloClient.getRepliesForComment(it1) }
```
Comment (leighdouglas) [inline app/src/main/java/com/kickstarter/viewmodels/ThreadViewModel.kt:72]: nit: here as well, can we rename `it1` to `comment` instead?

File: app/src/test/java/com/kickstarter/viewmodels/ThreadViewModelTest.kt [hunk 4]
```diff
@@ -27,7 +28,7 @@ class ThreadViewModelTest : KSRobolectricTestCase() {
     private val focusCompose = TestSubscriber<Boolean>()
     private val onReplies = TestSubscriber<List<CommentCardData>>()
 
-    private val createdAt = DateTime.now()
+    // private val createdAt = DateTime.now()
```
Comment (leighdouglas) [inline app/src/test/java/com/kickstarter/viewmodels/ThreadViewModelTest.kt:31]: is there a reason this is commented out? if we don't need it, it should be removed.

**Review by leighdouglas — APPROVED** [review]
Left a couple nitpick comments, otherwise code looks great!

**Review by Arkariang — APPROVED** [review]

**Pull request merged**